File: src/main/java/fi/quanfoxes/assembler/builders/References.java
package fi.quanfoxes.assembler.builders;

import fi.quanfoxes.assembler.*;
import fi.quanfoxes.assembler.references.*;
import fi.quanfoxes.parser.Node;
import fi.quanfoxes.parser.Variable;
import fi.quanfoxes.parser.nodes.ContentNode;
import fi.quanfoxes.parser.nodes.FunctionNode;
import fi.quanfoxes.parser.nodes.LinkNode;
import fi.quanfoxes.parser.nodes.NumberNode;
import fi.quanfoxes.parser.nodes.StringNode;
import fi.quanfoxes.parser.nodes.VariableNode;

public class References {
    public static enum ReferenceType {
        DIRECT,
        READ,
        VALUE,
        REGISTER
    }

    public static MemoryReference getObjectPointer(Unit unit) {
        return MemoryReference.parameter(unit, 0, 4);
    }

    private static Instructions getFunctionReference(Unit unit, FunctionNode node, ReferenceType type) {

        if (type == ReferenceType.DIRECT) {
            System.err.println("ERROR: Writable function return values aren't supported");
        }
        
        return Call.build(unit, node);
    }

    private static Instructions getLinkReference(Unit unit, LinkNode node, ReferenceType type) {
        return Link.build(unit, node, type);
    }

    /**
     * Returns a reference to a variable
     * @param node Variable as node
     * @param type Type of reference to get
     * @return Reference to a variable
     */
    private static Instructions getVariableReference(Unit unit, VariableNode node, ReferenceType type) {
        Variable variable = node.getVariable();

        if (type == ReferenceType.DIRECT) {
            unit.reset(variable);
        }
        else {
            Reference cache = unit.cached(variable);

            if (cache != null) {

                // Variable may be cached in stack
                if (!cache.isRegister()) {
                    Instructions move = Memory.toRegister(unit, cache);
                    return move.setReference(Value.getVariable(move.getReference(), variable));
                }

                return Instructions.reference(cache);
            }
        }

        Instructions instructions = new Instructions();
        Reference reference;

        switch (variable.getVariableType()) {
            
            case GLOBAL: {
                reference = ManualReference.global(variable.getFullname(), Size.get(variable.getType().getSize()));
                break;
            }

            case LOCAL: {
                reference = MemoryReference.local(unit, variable.getAlignment(), variable.getType().getSize());
                break;
            }

            case PARAMETER: {
                reference = MemoryReference.parameter(unit, variable.getAlignment(), variable.getType().getSize());
                break;
            }

            case MEMBER: {
                Register register = unit.getObjectPointer();

                if (register == null) {
                    Instructions move = Memory.getObjectPointer(unit, type);
                    instructions.append(move);

                    register = move.getReference().getRegister();
                }

                reference = MemoryReference.member(register, variable.getAlignment(), variable.getType().getSize());
                break;
            }

            default: {
                return null;
            }
        }

        if ((type == ReferenceType.VALUE || type == ReferenceType.REGISTER) && reference.isComplex()) {
            Instructions move = Memory.toRegister(unit, reference);
            instructions.append(move);

            return instructions.setReference(Value.getVariable(move.getReference(), variable));
        }
        
        return instructions.setReference(reference);
    }

    private static Instructions getNumberReference(Unit unit, NumberNode node, ReferenceType type) {
        Size size = Size.get(node.getType().getSize());

        switch (type) {
            case DIRECT:
                return Instructions.reference(new AddressReference(node.getValue()));
            case VALUE:
            case READ:
                return Instructions.reference(new NumberReference(node.getValue(), size));
            case REGISTER:
                Instructions instructions = Memory.toRegister(unit, new NumberReference(node.getValue(), size));
                instructions.setReference(Value.getNumber(instructions.getReference().getRegister(), size));
                return instructions;
        }

        return null;
    }

    private static Instructions getStringReference(Unit unit, StringNode node, ReferenceType type) {
        switch (type) {
            
            case DIRECT: {
                return Instructions.reference(ManualReference.string(node.getIdentifier(), true));
            }

            case READ: {
                return Instructions.reference(ManualReference.string(node.getIdentifier(), false));
            }

            case VALUE:
            case REGISTER: {
                Reference reference = ManualReference.string(node.getIdentifier(), false);
                Instructions instructions = Memory.toRegister(unit, reference);

                return instructions.setReference(Value.getString(instructions.getReference().getRegister()));
            }

            default: return null;
        }
    }

    public static Instructions get(Unit unit, Node node, ReferenceType type) {
        if (node instanceof FunctionNode) {
            return getFunctionReference(unit, (FunctionNode)node, type);
        }
        else if (node instanceof LinkNode) {
            return getLinkReference(unit, (LinkNode)node, type);
        }
        else if (node instanceof VariableNode) {
            return getVariableReference(unit, (VariableNode)node, type);
        }
        else if (node instanceof NumberNode) {
            return getNumberReference(unit, (NumberNode)node, type);
        }
        else if (node instanceof ContentNode) {
            return get(unit, node.first(), type);
        }
        else if (node instanceof StringNode) {
            return getStringReference(unit, (StringNode)node, type);
        }
        else {

            if (type == ReferenceType.DIRECT) {
                System.out.println("Warning: Complex writable reference requested");
            }

            return unit.assemble(node);
        }
    }

    public static Instructions direct(Unit unit, Node node) {
        return References.get(unit, node, ReferenceType.DIRECT);
    }

    public static Instructions read(Unit unit, Node node) {
        return References.get(unit, node, ReferenceType.READ);
    }

    public static Instructions value(Unit unit, Node node) {
        return References.get(unit, node, ReferenceType.VALUE);
    }

    public static Instructions register(Unit unit, Node node) {
        return References.get(unit, node, ReferenceType.REGISTER);
    }

    /**
     * Returns whether node is primitive (Requires only building references)
     */
    private static boolean isPrimitive(Node node) {
        return node instanceof VariableNode || node instanceof StringNode || node instanceof NumberNode;
    }

    /**
     * Returns references to both given nodes
     * @param program Program to append the instructions for referencing the nodes
     * @param a Node a
     * @param b Node b
     * @param at Node a's reference type
     * @param bt Node a'b reference type
     * @return References to both of the nodes
     */
    public static Reference[] get(Unit unit, Instructions program, Node a, Node b, ReferenceType at, ReferenceType bt) {
        Reference[] references = new Reference[2];

        if (!isPrimitive(a)) {
            Instructions instructions = References.get(unit, a, at);
            references[0] = instructions.getReference();

            program.append(instructions);
        }

        if (!isPrimitive(b)) {
            Instructions instructions = References.get(unit, b, bt);
            references[1] = instructions.getReference();

            program.append(instructions);
        }

        if (references[0] == null) {
            Instructions instructions = References.get(unit, a, at);
            references[0] = instructions.getReference();

            program.append(instructions);
        }
        
        if (references[1] == null) {
            Instructions instructions = References.get(unit, b, bt);
            references[1] = instructions.getReference();

            program.append(instructions);
        }

        return references;
    }
}

File: Zigzag/Assembler/Builders/References.cs
using System;
using System.Linq;

public enum ReferenceType
{
	DIRECT,
	READ,
	VALUE,
	REGISTER,
	DEFAULT
}

public class References
{
	public static MemoryReference GetObjectPointer(Unit unit)
	{
		return MemoryReference.Parameter(unit, 0, 4);
	}

	private static Instructions GetFunctionReference(Unit unit, FunctionNode node, ReferenceType type)
	{
		if (type == ReferenceType.DIRECT)
		{
			Console.Error.WriteLine("ERROR: Writable function return values aren't supported");
		}

		return Call.Build(unit, node);
	}

	private static Instructions GetLinkReference(Unit unit, LinkNode node, ReferenceType type)
	{
		return Links.Build(unit, node, type);
	}

	private static Instructions GetVariableReference(Unit unit, VariableNode node, ReferenceType type)
	{
		return GetVariableReference(unit, node.Variable, type);
	}

	public static Instructions GetVariableReference(Unit unit, Variable variable, ReferenceType type)
	{
		if (type == ReferenceType.DIRECT)
		{
			unit.Reset(variable);
		}
		else
		{
			var cache = unit.IsCached(variable);

			if (cache != null)
			{
				// Variable may be cached in stack
				if (!cache.IsRegister())
				{
					var move = Memory.ToRegister(unit, cache);
					return move.SetReference(global::Value.GetVariable(move.Reference, variable));
				}

				return Instructions.GetReference(cache);
			}
		}

		var instructions = new Instructions();

		Reference reference;

		switch (variable.Category)
		{
			case VariableCategory.GLOBAL:
			{
				reference = ManualReference.Global(variable.Name, Size.Get(variable.Type.Size));
				break;
			}

			case VariableCategory.LOCAL:
			{
				reference = MemoryReference.Local(unit, variable.Alignment, variable.Type.Size);
				break;
			}

			case VariableCategory.PARAMETER:
			{
				reference = MemoryReference.Parameter(unit, variable.Alignment, variable.Type.Size);
				break;
			}

			case VariableCategory.MEMBER:
			{
				var register = unit.GetObjectPointer();

				if (register == null)
				{
					var move = Memory.GetObjectPointer(unit, type);
					instructions.Append(move);

					register = move.Reference.GetRegister();
				}

				reference = MemoryReference.Member(register, variable.Alignment, variable.Type.Size);
				break;
			}

			default:
			{
				return null;
			}
		}

		reference.Metadata = variable;

		if ((type == ReferenceType.VALUE || type == ReferenceType.REGISTER) && reference.IsComplex())
		{
			var move = Memory.ToRegister(unit, reference);
			instructions.Append(move);

			return instructions.SetReference(global::Value.GetVariable(move.Reference, variable));
		}

		return instructions.SetReference(reference);
	}

	private static Instructions GetNumberReference(Unit unit, NumberNode node, ReferenceType type)
	{
		var size = Size.Get(node.Type.Size);

		switch (type)
		{
			case ReferenceType.DIRECT: return Instructions.GetReference(new AddressReference(node.Value));

			case ReferenceType.VALUE:
			case ReferenceType.READ: return Instructions.GetReference(new NumberReference(node.Value, size));

			case ReferenceType.REGISTER:
			{
				Instructions instructions = Memory.ToRegister(unit, new NumberReference(node.Value, size));
				instructions.SetReference(global::Value.GetNumber(instructions.Reference.GetRegister(), size));
				return instructions;
			}

		}

		return null;
	}

	private static Instructions GetStringReference(Unit unit, StringNode node, ReferenceType type)
	{
		switch (type)
		{
			case ReferenceType.DIRECT:
			{
				return Instructions.GetReference(ManualReference.String(node.Identifier, true));
			}

			case ReferenceType.READ:
			{
				return Instructions.GetReference(ManualReference.String(node.Identifier, false));
			}

			case ReferenceType.VALUE:
			case ReferenceType.REGISTER:
			{
				var reference = ManualReference.String(node.Identifier, false);
				var instructions = Memory.ToRegister(unit, reference);

				return instructions.SetReference(global::Value.GetString(instructions.Reference.GetRegister()));
			}

			default: return null;
		}
	}

	public static Instructions Get(Unit unit, Node node, ReferenceType type)
	{
		switch (node.GetNodeType())
		{
			case NodeType.FUNCTION_NODE:
			{
				return GetFunctionReference(unit, (FunctionNode)node, type);
			}

			case NodeType.LINK_NODE:
			{
				return GetLinkReference(unit, (LinkNode)node, type);
			}

			case NodeType.VARIABLE_NODE:
			{
				return GetVariableReference(unit, (VariableNode)node, type);
			}

			case NodeType.NUMBER_NODE:
			{
				return GetNumberReference(unit, (NumberNode)node, type);
			}

			case NodeType.CONTENT_NODE:
			{
				return References.Get(unit, node.First, type);
			}

			case NodeType.STRING_NODE:
			{
				return GetStringReference(unit, (StringNode)node, type);
			}

			case NodeType.CAST_NODE:
			{
				return References.Get(unit, node.First, type);
			}

			case NodeType.OPERATOR_NODE:
			{
				var operation = node as OperatorNode;

				if (operation.Operator == Operators.EXTENDER)
				{
					return Arrays.Build(unit, operation, type);
				}

				if (type == ReferenceType.DIRECT)
				{
					Console.WriteLine("Warning: Complex writable reference requested");
				}

				return unit.Assemble(node, type);
			}

			default:
			{
				if (type == ReferenceType.DIRECT)
				{
					Console.WriteLine("Warning: Complex writable reference requested");
				}

				return unit.Assemble(node);
			}
		}
	}

	public static Instructions Direct(Unit unit, Node node)
	{
		return References.Get(unit, node, ReferenceType.DIRECT);
	}

	public static Instructions Read(Unit unit, Node node)
	{
		return References.Get(unit, node, ReferenceType.READ);
	}

	public static Instructions Value(Unit unit, Node node)
	{
		return References.Get(unit, node, ReferenceType.VALUE);
	}

	public static Instructions Register(Unit unit, Node node)
	{
		return References.Get(unit, node, ReferenceType.REGISTER);
	}

	/**
     * Returns whether node is primitive (Requires only building references)
     */
	private static bool IsPrimitive(Node node)
	{
		return node.GetNodeType() == NodeType.VARIABLE_NODE || node.GetNodeType() == NodeType.STRING_NODE || node.GetNodeType() == NodeType.NUMBER_NODE;
	}

	private const int FUNCTION_WEIGHT = 2;
	private const int DEFAULT_WEIGHT = 1;

	private static int GetWeight(Node node)
	{
		return node.Select(n =>
		{
			switch (n.GetNodeType())
			{
				case NodeType.CONSTRUCTION_NODE: return FUNCTION_WEIGHT;
				case NodeType.FUNCTION_NODE: return FUNCTION_WEIGHT;
				case NodeType.INLINE_NODE: return FUNCTION_WEIGHT;

				default: return DEFAULT_WEIGHT;
			}
		}).Sum();
	}

	/// <summary>
	/// Returns references to the nodes
	/// </summary>
	/// <param name="unit">Unit used to operate</param>
	/// <param name="program">Instructions where the operations are appended</param>
	/// <param name="n1">First node</param>
	/// <param name="n2">Second node</param>
	/// <param name="r1t">Type of reference to get for the first node</param>
	/// <param name="r2t">Type of reference to get for the second node</param>
	/// <param name="r1">Result: Reference for the first node</param>
	/// <param name="r2">Result: Reference for the second node</param>
	public static void Get(Unit unit, Instructions program, Node n1, Node n2, ReferenceType r1t, ReferenceType r2t, out Reference r1, out Reference r2)
	{
		r1 = null;
		r2 = null;

		var w1 = GetWeight(n1);
		var w2 = GetWeight(n2);

		if (w1 >= w2)
		{
			var instructions = References.Get(unit, n1, r1t);
			r1 = instructions.Reference;

			program.Append(instructions);

			instructions = References.Get(unit, n2, r2t);
			r2 = instructions.Reference;

			program.Append(instructions);
		}
		else
		{
			var instructions = References.Get(unit, n2, r2t);
			r2 = instructions.Reference;

			program.Append(instructions);

			instructions = References.Get(unit, n1, r1t);
			r1 = instructions.Reference;

			program.Append(instructions);
		}
	}

	/**
     * Returns references to both given nodes
     * @param program Program to append the instructions for referencing the nodes
     * @param a Node a
     * @param b Node b
     * @param at Node a's reference type
     * @param bt Node a'b reference type
     * @return References to both of the nodes
     */
	public static Reference[] Get(Unit unit, Instructions program, Node a, Node b, ReferenceType at, ReferenceType bt)
	{
		var references = new Reference[2];

		if (!IsPrimitive(a))
		{
			var instructions = References.Get(unit, a, at);
			references[0] = instructions.Reference;

			program.Append(instructions);
		}

		if (!IsPrimitive(b))
		{
			var instructions = References.Get(unit, b, bt);
			references[1] = instructions.Reference;

			program.Append(instructions);
		}

		if (references[0] == null)
		{
			var instructions = References.Get(unit, a, at);
			references[0] = instructions.Reference;

			program.Append(instructions);
		}

		if (references[1] == null)
		{
			var instructions = References.Get(unit, b, bt);
			references[1] = instructions.Reference;

			program.Append(instructions);
		}

		return references;
	}
}